design/aws-ecs/aws-ecs-tagging-support.md

```diff
@@ -0,0 +1,196 @@
+Amazon ECS supports tagging for all the ECS resources (see [here](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-using-tags.html)). Note that ECS provides a property of ECS service named `PropagateTags` to support propagating tags from either task definitions or services to tasks (tasks cannot be tagged using CDK without setting this parameter). Currently this property is not set in CDK so that the tags are not propagated by default.
+
+In addition, `EnableECSManagedTags` is provided as another property so that ECS managed tags can be supported and used to tag ECS tasks. However, currently users cannot set `EnableECSManagedTags` and the default value is `false`.
+
+Additionally, CDK support tagging and can cascade tags to all its tagable children (see [here](https://docs.aws.amazon.com/cdk/latest/guide/tagging.html)). The current CDK tagging API is shown below:
+
+``` ts
+myConstruct.node.applyAspect(new Tag('key', 'value'));
+
+myConstruct.node.applyAspect(new RemoveTag('key', 'value'));
+```
+
+As we can see, the current tagging API is not nice and intuitive enough for users which needs to be improved.
+
+## General approach
+
+The new `BaseService` class will include two more base properties:
+
+* propagateTags
+* enableECSManagedTags
+
+`propagateTags` specifies whether to propagate the tags from the task definition or the service to ECS tasks and `enableECSManagedTags` specifies whether to enable Amazon ECS managed tags for the tasks within the service. Also, for `Ec2Service` and `FargateService` we will have the corresponding two new properties exposed to users:
+
+* propagateTaskTagsFrom
+* enableECSManagedTags
+
+*Note that the reason why we define `propagateTaskTagsFrom` is because we want to have a different name for `propagateTags` to eliminate the naming confusion.*
+
+For the tagging behavior part, we have three options for the new tagging API:
+
+**Option 1: define public static method**
+
+``` ts
+Tag.apply(myConstruct, 'key', 'value');
+
+RemoveTag.apply(myConstruct, 'key');
+```
+
+**Option 2: use Tag object (RECOMMENDED since users can reuse tags)**
+
+``` ts
+const tag = new Tag('key', 'value');
+tag.apply(myConstruct);
+
+const removetag = new RemoveTag('key');
+removetag.apply(myConstruct);
+```
+
+**Option 3: define method in Construct**
+
+``` ts
+myConstruct.apply(new Tag('key', 'value'));
+
+myConstruct.apply(new RemoveTag('key'));
+```
+
+## Code changes
+
+Given the above, we should make the following changes on ECS:
+* Support tag propagation for ECS task and ECS managed tags
+  1. Add `propagateTags` and `enableECSManagedTags` properties to `BaseServiceOptions`.
+  2. Add `propagateTaskTagsFrom` and `enableECSManagedTags` properties to `Ec2ServiceProps` and `FargateServiceProps`.
+  3. Add an enum type `PropagateTagsFromType` to support `propagateTaskTagsFrom`.
+* Change CDK Tagging API
+  1. Add a method `apply(scope, 'key', 'value')` to `Tag` class and `RemoveTag` class, which calls `applyAspect`.
```

Comment:
I don't think this is necessary in any way, and I don't like the proposed APIs any more than I like the current API. In particular, this is very thin sugar on top of just using `applyAspec`, which is additional API surface to maintain for a very minimal value.

Reply:
Fixed!